### PR TITLE
Fletcher32: Add incremental API

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR extends the current fletcher32 checksum with an incremental API mode. This way the bytes to be checksummed can be supplied via multiple successive calls and do not have to be provided in a single consecutive buffer.

I've also rephrased the warning with the original function a bit as that function uses an `unaligned_get_u16` to access the data. The data thus does not require alignment, but the length does need to be supplied as number of 16 bit words.

### Testing procedure

The test has been extended

### Issues/PRs references

None

### CODE REVIEW

File: sys/include/checksum/fletcher32.h [hunk 1]
```diff
@@ -27,22 +28,59 @@
...
+ * @brief Initialize a fletcher32 context
+ *
+ * Multi-part version of @ref fletcher32.
+ *
+ * @param[in]   ctx     fletcher32 context to initialize
+ */
+void fletcher32_init(fletcher32_ctx_t *ctx);
+
+/**
+ * @brief Update the fletcher32 context with new data
```
Comment (Teufelchen1) [inline sys/include/checksum/fletcher32.h:64]: How often can I update the context?

File: sys/checksum/fletcher32.c [hunk 2]
```diff
@@ -21,21 +21,42 @@
...
+
+void fletcher32_update(fletcher32_ctx_t *ctx, const uint8_t *data, size_t words)
+{
+    const uint16_t *u16_data = (const uint16_t*)data;
     while (words) {
         unsigned tlen = words > 359 ? 359 : words;
         words -= tlen;
         do {
-            sum2 += sum1 += unaligned_get_u16(data++);
+            ctx->sum2 += ctx->sum1 += unaligned_get_u16(u16_data++);
```
Comment (Teufelchen1) [inline sys/checksum/fletcher32.c:50]: Nit: Hard to read & comprehend for humans who are new to C. Maybe split into 2..3 lines?

File: sys/checksum/fletcher32.c [hunk 3]
```diff
@@ -21,21 +21,42 @@
...
+uint32_t fletcher32_finish(fletcher32_ctx_t *ctx)
+{
+    /* Second reduction step to reduce sums to 8 bits */
+    _reduce(ctx);
+    return (ctx->sum2 << 16) | ctx->sum1;
+}
+
+void fletcher32_update(fletcher32_ctx_t *ctx, const uint8_t *data, size_t words)
+{
+    const uint16_t *u16_data = (const uint16_t*)data;
```
Comment (Teufelchen1) [inline sys/checksum/fletcher32.c:45]: Maybe assert that the pointers are not `NULL`?

File: sys/include/checksum/fletcher32.h [hunk 4]
```diff
@@ -27,22 +28,59 @@
...
+ * @note @p data should contain an integer number of 16 bit words
+ *
+ * @param[in]   ctx     fletcher32 context
+ * @param[in]   data    Data to add to the context
+ * @param[in]   words   Length of the data in 16 bit words
+ */
+void fletcher32_update(fletcher32_ctx_t *ctx, const uint8_t *data, size_t words);
+
+/**
+ * @brief Finalize the checksum operation and return the checksum
```
Comment (Teufelchen1) [inline sys/include/checksum/fletcher32.h:76]: In the doc of `fletcher32(..)` it is noted that the sum is never zero. Isn't that also true here?

**Review by Teufelchen1 — CHANGES_REQUESTED** [review]
Only minor remarks by me and one static-test! :muscle:

File: sys/checksum/fletcher32.c (hunk 3 above)
Comment (bergzand) [reply sys/checksum/fletcher32.c:45]: `NULL` is fine here as long as `words` is also zero, `fletcher32_update(ctx, NULL, 0);` should be valid.

File: sys/include/checksum/fletcher32.h (hunk 1 above)
Comment (bergzand) [reply sys/include/checksum/fletcher32.h]: Added clarification here

File: sys/checksum/fletcher32.c (hunk 2 above)
Comment (bergzand) [reply sys/checksum/fletcher32.c]: Fair, split into two lines

File: sys/include/checksum/fletcher32.h (hunk 4 above)
Comment (bergzand) [reply sys/include/checksum/fletcher32.h:77]: Added `@return` documentation with this.

**Review by Teufelchen1 — APPROVED** [review]
:shrimp:

**Comment by kaspar030** [comment]
... waiting for users passing in byte array length as "words" ...

**Comment by bergzand** [comment]
> ... waiting for users passing in byte array length as "words" ...

The whole api of this module has in my opinion always been a huge footgun. Any suggestions here?

**Comment by kaspar030** [comment]
> The whole api of this module has in my opinion always been a huge footgun. Any suggestions here?

Agreed. Now that there's a struct with state, buffer single bytes there if necessary?

**Comment by kaspar030** [comment]
... but of course I don't want to hold up the already ACK'ed PR.

File: sys/include/checksum/fletcher32.h [hunk 5]
```diff
@@ -27,22 +28,60 @@
...
+ * times with new data to checksum.
+ *
+ * @note @p words is the number of 16 bit words in the buffer
+ * @note @p data should contain an integer number of 16 bit words
+ *
+ * @param[in]   ctx     fletcher32 context
+ * @param[in]   data    Data to add to the context
+ * @param[in]   words   Length of the data in 16 bit words
+ */
+void fletcher32_update(fletcher32_ctx_t *ctx, const uint8_t *data, size_t words);
```
Comment (benpicco) [inline sys/include/checksum/fletcher32.h:74]: might as well make it
```suggestion
void fletcher32_update(fletcher32_ctx_t *ctx, const void *data, size_t words);
```

**Comment by bergzand** [comment]
> Agreed. Now that there's a struct with state, buffer single bytes there if necessary?

1. Follow up
2. How would you handle the case where the full length of the (incrementally passed) data is odd?

File: sys/include/checksum/fletcher32.h (hunk 5 above)
Comment (bergzand) [reply sys/include/checksum/fletcher32.h:74]: Fixed!

File: sys/checksum/fletcher32.c [hunk 6]
```diff
@@ -21,21 +21,43 @@
...
+uint32_t fletcher32_finish(fletcher32_ctx_t *ctx)
+{
+    /* Second reduction step to reduce sums to 8 bits */
+    _reduce(ctx);
+    return (ctx->sum2 << 16) | ctx->sum1;
+}
+
+void fletcher32_update(fletcher32_ctx_t *ctx, const void *data, size_t words)
+{
+    const uint16_t *u16_data = (const uint16_t*)data;
```
Comment (benpicco) [inline sys/checksum/fletcher32.c:45]:
```suggestion
    const uint16_t *u16_data = data;
```
no need to cast the `void *`

Comment (bergzand) [reply sys/checksum/fletcher32.c:45]: Ugh 😳

**Comment by benpicco** [comment]
Feel free to squash

**Comment by bergzand** [comment]
Squashed!

**Pull request merged**